tchannel-core/src/test/java/com/uber/tchannel/handlers/LoadControlHandlerTest.java

```diff
@@ -0,0 +1,173 @@
+package com.uber.tchannel.handlers;
+
+import static org.junit.Assert.assertEquals;
+
+import com.uber.tchannel.api.ResponseCode;
+import com.uber.tchannel.api.SubChannel;
+import com.uber.tchannel.api.TChannel;
+import com.uber.tchannel.api.TChannel.Builder;
+import com.uber.tchannel.api.TFuture;
+import com.uber.tchannel.api.handlers.RequestHandler;
+import com.uber.tchannel.errors.ErrorType;
+import com.uber.tchannel.messages.RawRequest;
+import com.uber.tchannel.messages.RawResponse;
+import com.uber.tchannel.messages.Request;
+import com.uber.tchannel.messages.Response;
+import java.net.InetAddress;
+import java.util.concurrent.Semaphore;
+import org.junit.Test;
+
+public class LoadControlHandlerTest {
+
+    @Test(expected = IllegalArgumentException.class)
+    public void invariant_lowLessThanZero() {
+        new Builder("server")
+            .setChildLoadControl(-1, 10)
+            .build();
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void invariant_highEqualToLow() {
+        new TChannel.Builder("server")
+            .setChildLoadControl(10, 10)
+            .build();
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void invariant_highLessThanLow() {
+        new TChannel.Builder("server")
+            .setChildLoadControl(10, 5)
+            .build();
+    }
+
+    /**
+     * Send two requests in parallel.
+     *
+     * One of the requests will not be resolved, because it triggers the high water mark.
+     *
+     * The request that _does_ resolve, will not trigger the low water mark.
+     *
+     * So the other request will timeout.
+     */
+    @Test
+    public void testHighWaterMark() throws Exception {
+
+        try (Scenario scenario = new Scenario(0, 2)) {
```

Comment:
If I understand the code/test scenario correctly, the first message goes through, whereas the second one triggers the high watermark. If by that time the second message is already read, why does it time out?

Reply:
Both requests are read.
Both requests reach the LoadControlHandler.
One request increments `outstanding` to 1.
This request goes on to the handler.
The other request increments `outstanding` to 2; high water mark is breached and reading is paused.
Because auto-read is paused, this request does not go on to the handler.
Auto-read is never resumed so this request times out.

Now I actually realize this is not the best, and we should allow the 'triggering' request to move on.

Reply:
Please see the revised test